### PR TITLE
[#71] Using @schema_prefix breaks prefix option

### PR DESCRIPTION
**Pull request by maltoe**
[[#71]](https://github.com/bitcrowd/carbonite/issues/71)

### CODE REVIEW

**Comment by maltoe** [comment]
@andreasknoepfle As mentioned in the ticket, it is possible to override the `@schema_prefix` by setting a `prefix` option on the `from`, only it has to be a string at compile time:

```elixir
iex(1)> from(t in Trigger, prefix: ^carbonite_prefix) |> Repo.all()

** (Ecto.Query.CompileError) `prefix` must be a compile time string, got: `{:ok, ^carbonite_prefix}`
```

However, I did manage to make it work by circumventing this restriction and manually updating the query:

```elixir
iex(2)> from(t in Trigger) |> then(fn q -> Map.put(q, :from, Map.put(q.from, :prefix, carbonite_prefix)) end) -> Repo.all()
...
```

Wdyt? Feels like cheating right? Also of course using the internals of the `Ecto.Query` struct.

File: lib/carbonite/prefix.ex [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    quote do
+      unquote(queryable)
+      |> Ecto.Query.join(unquote(type), unquote(binding), unquote(join_expr))
+      |> apply_carbonite_prefix_on_join_expr(unquote(opts))
+    end
+  end
+
+  @doc false
+  @spec apply_carbonite_prefix_on_from_expr(Ecto.Query.t(), keyword) :: Ecto.Query.t()
+  def apply_carbonite_prefix_on_from_expr(%{from: from} = queryable, opts) do
```
Comment (maltoe) [inline lib/carbonite/prefix.ex:34]: can be private

File: lib/carbonite/prefix.ex [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    # Inject the prefix into the FromExpr manually.
+    %{queryable | from: apply_carbonite_prefix(from, opts)}
+  end
+
+  def apply_carbonite_prefix_on_join_expr(%{joins: [join]} = queryable, opts) do
```
Comment (maltoe) [inline lib/carbonite/prefix.ex:39]: `@spec` `@doc false`

**Review by andreasknoepfle — APPROVED** [review]
👏

**Pull request merged**